### ISSUE TITLE
Skip boost if there is a local/valid Installation. Force boost bld /B

### ISSUE THREAD

**Issue by jasin** [geoffmcl/test-sg#11]
Feature request
I would like the option to avoid building boost if I already have a valid installation. 
1. system environment %BOOST_ROOT% contains a valid path to a boost installation
2. Validate %BOOST_ROOT% - thoughts on this? Other then just %BOOST_ROOT% != NUL
3. /B - Use built-in Boost, I propose this switch

**Comment by geoffmcl** [comment]
@jasin I am unsure what you want the /B switch to do? What does `Use built-in Boost` mean?

Yes, I can see an idea that if BOOST_ROOT, Boost_DIR, or BOOSTROOT, is set to a valid boost installation, before the make3rd.x64 is run, then the :SET_BOOST function could return that... and no boost download or clone would be done... Is that what you are talking about?

I certainly agree there is presently a problem with using the simpits binaries... I would prefer to do a clone, and if necessary a full library build... as d-and-c.x64 did previously...

So what is your idea here?

**Comment by jasin** [comment]
CMake runs and tries to locate boost in the usual area's. One of those it checks is if there is a system env. variable set, namely %BOOST_ROOT%. 

If %BOOST_ROOT% is already set by a system env. variable then I want to skip building boost in make3rd.x64.bat because I already have a valid d-and-c of boost locally, just in a different directory.

The /B switch, or whatever we call it, will tell the script regardless of what %BOOST_ROOT% is prior to running the script, build boost locally and set %BOOST_ROOT% to use the d-and-c in the make3rd.x64.bat

Now my rational for such a feature. It would not affect any user wanting to solely d-and-c SG/FG. They need boost, don't have boost, and will get boost when script is run. 
I, for example, have boost installed. I don't want to waste my time d-and-c in make3rd.x64.bat when I have a valid boost install ready to use and cmake knows where it is at.

Now give the user freedom, make a switch /B to force the build in the %WORKSPACE% directory and use that boost version instead by setting %BOOST_ROOT% to %WORKSPACE%\Boost while script executes. 

 the sole purpose is to avoid duplicating efforts by skipping valid installed software that can be used. If you don't have them, we build them regardless

**Comment by jasin** [comment]
or maybe even better would be if the switch just allowed me to avoid building boost. Either or. I guess it could be assumed if i already have boost then I should know enough to disable it. Would that better? Don't want to overthink this..

**Comment by geoffmcl** [comment]
@jasin all understood... the priorities are -
1. build cleanly for a new user, that has never heard of boost
2. add advanced user features

So, as explained I am not happy with the current boost choice, so adding a /B switch, is, well a thought...

As you point out, an advanced user can just do a few quick patches, and the scripts do what [s]he wants... and if their patch also works for new users, then they could suggest it...

I want to move the Boost stuff out of d-and-c, and only have it in make3rd. Then a /B switch could toggle between simpits binary or git clone compile...

And when this is all correctly in **one** place, then it would be easy to add an advanced feature... 

If the User sets `Boost_DIR` **before** running the scripts, then make3rd would skip any type of Boost download, and use the boost installation provided... This should **not** require a command line switch...

Does that fit into your `rational`?

It seems we want the same things...

**Comment by geoffmcl** [comment]
@jasin if you missed this elsewhere, have merge **all** back to `master`... you should rebase to there, and branch from it, for development... thanks...

**Comment by jasin** [comment]
@geoffmcl 
Yes I believe that should work and fits exactly what I want to do. I just need to define %BOOST_DIR% before running script. Perfect

**Comment by jasin** [comment]
@geoffmcl 
I was doing a dry run and letting the script download boost inside the script directory. **BIG PROBLEM** The url to download boost is pre-compiled for MSVC 10, as such cmake can't find thread and system lib's and fails at CGAL, the first build following Boost. 

example
`libboost_thread-vc100-mt-1_55`
needs to be
`libboost_thread-vc120-mt-1_5x`

x = some number greater then 6 and I had problems with 1.5.6 on MSVC 12, but 1.5.5 is definitely broken. I am using, successfully, boost 1.6.0 to date

Is it possible to also add a d-and-c here if MSVC > 10?

edit: Or assume MSVC 12 users install and compile boost **prior** to running make3rd.x64.bat and 
`@set BOOST_DIR=<*\boost>`

**Comment by geoffmcl** [comment]
@jasin yes, the problems you point out are known...
1. As stated, I want to move the boost stuff out of d-and-c, and put all boost in make3rd... and make the svn download and compile of boost the default...
2. Then provide a way the boost root is set for when d-and-c runs after make3rd...
3. With that done, working and tested, we can give some consideration to alternatively using a previous install of boost... base on environment variables...
4. Add a /? option switch that used simpits instead of svn, if that is needed...

But it is the order that is important... 1 and 2 must be done before 3 can be a reality ;=))

Any assistance appreciated...

Step by step...
